Graphs linechart

steviekaye opened the pull request:
![budget-tracker_stackedcolumnchart](https://user-images.githubusercontent.com/18640469/52157842-36391900-26f7-11e9-875c-ac7cd0282033.png)

Added a stacked column chart to display category breakdown by months. Includes significant data sorting in the Summaries controller to massage the data into a shape suitable for ChartKick, and some initial/incomplete testing. Could be modified to show e.g. last six months by default. Currently doesn't have a title while I think of the default range.

steviekaye committed:
Preliminary work on data grouping for line charts

steviekaye committed:
Merge branch 'master' of https://github.com/steviekaye/budget-tracker into graphs-linechart

steviekaye committed:
Stacked column chart initial implementation

steviekaye committed:
Tweaked stacked bar chart to provide values "empty" months.
Added pry to gemfile for better testing.
Added specs for shape of data.

steviekaye committed:
Fixed typo

kfrn commented on app/controllers/summaries_controller.rb:
```diff
@@ -15,7 +16,42 @@ def category_totals
+      contained_months = c[:data].map(&:first)
+      blank_months = date_range - contained_months
+      blank_months.each { |b| insert_empty_month(c[:data], [b, 0]) }
+    end
```
This is pretty dense and to be honest I haven't got to grips with it totally, it will be easier once the other naming issues are sorted out. Just wondering if it's possible to replace this FN with something like this (not proper code):
```ruby
def sum_per_category_and_month(purchases_by_category, all_months)
    purchases_by_category.map do |cat|
      summed = # some code
      # Result like:
      #  { name: 'Basics', data: [['November 2018', 100], ['December 2018', 150]] }

      zero_purchase_months = # months that are in all_months, but not summer.data
      # Add in months with total of zero:
      ?zero_purchase_months.each { |m| summed.data.push({[m, 0]}) }

      summed
    end
  end
```

Some of that is similar to what you have already, I'm just thinking it through in the writing.

steviekaye replied:
What the existing method offers which this alternative doesn't is maintaining month order - will take another look at refactoring once other concerns are addressed.

kfrn reviewed (changes requested):
Looks good. I think there are some improvements to be made, especially around accurate naming - see the notes.

steviekaye committed:
More accurate naming for summary data methods
Shifted some code from blocks to inline

steviekaye committed:
Further renaming and refactoring of methods

steviekaye committed:
Refactored data sorting:
 - Moved from Summaries Controller to new service class
 - Updated specs to match

steviekaye committed:
Changed number format in tests to be human-readable

kfrn commented on app/services/data_formatter.rb:
```diff
@@ -0,0 +1,57 @@
+      end }
+    end
+
+    summaries.each do |c|
```
Also, I think I'd extract this block into a new method, e.g. `ensure_all_months_present`.
`first_purchase`, `last_purchase`, and `months_in_range` can actually just be defined in there, rather than passed down from `category_totals_by_month`.

kfrn replied:
I was playing around with this, here's the implementation I came up with:

```ruby
  def ensure_all_months_present(summaries)
    first_purchase = Purchase.all.order(:date).first[:date] # TODO: better query
    last_purchase = Purchase.all.order(:date).last[:date]
    months_in_range = (first_purchase..last_purchase).map { |d| d.strftime('%B %Y') }.uniq

    summaries.each do |summary|
      month_totals = months_in_range.map do |month|
        existing_total = summary[:data].detect { |s| s.first == month }

        existing_total ? existing_total : [month, 0]
      end

      summary[:data] = month_totals
    end
  end
```
Not sure if it is the 'best' way to tackle this, but possibly food for thought.

kfrn replied:
Oh sorry, I didn't mean that you had to do it that way! It was just what I came up with.

kfrn commented on app/services/data_formatter.rb:
```diff
@@ -0,0 +1,57 @@
+  end
+
+  def category_totals_by_month
+    first_purchase = Purchase.all.order(:date).first[:date]
```
Is there a way to just get the first purchase back from the DB? Currently you're getting all purchases from the DB, then ordering, then taking the first. It should be possible to write a query that just gets one answer, the first by date.

steviekaye replied:
Having had a play around, it turns out you can use .maximum / .minimum on Activerecords and it can order by date as well as returning just that field - much cleaner. Thanks!

kfrn reviewed (changes requested):
Loving the development of this! I've got a couple more suggestions, see notes above

kfrn commented on app/services/data_formatter.rb:
```diff
@@ -0,0 +1,57 @@
+  end
+
+  def extract_purchases_by_category(category)
+    @purchases.where(subcategory: category.subcategories).map { |p| { date: p.date, amount: p.amount } }
```
Actually, I'd say that this could now be a var inside the loop in `purchases_by_category`, rather than a separate method.

Also,in `purchases_by_category` when you loop over the purchases per category, you already have the `date` and `amount` k-v pairs, so you can probably just add in `name` rather than repeat the other two.

steviekaye replied:
Seems reasonable if you think that's not putting too much logic in the former.

And good suggestion - wasn't sure if it could be done, but .merge! seems to be the solution.

steviekaye committed:
Altered various methods in data_formatter
- consolidated some, refactored others.

steviekaye committed:
Further refactoring and simplification

kfrn commented on app/services/data_formatter.rb:
```diff
@@ -0,0 +1,56 @@
+  end
+
+  def purchases_by_category
+    totals = Category.all.map do |category|
```
Nitpick (sorry) - is `totals` an accurate name?

kfrn reviewed (approved):
Awesome work, this is looking really good! Interesting to compare the [before](https://github.com/steviekaye/budget-tracker/blob/e7e60ba31704138d517937a4e688f2828f628e0a/app/controllers/summaries_controller.rb) and [after](https://github.com/steviekaye/budget-tracker/blob/1d5fe96bf9be169502991917243f730bfd25cc9c/app/services/data_formatter.rb).

I'd say you're about ready to merge. Although the tests don't explicitly say so, you're already testing that the formatter is entering the 'zero' months.   
If you do decide to update the tests, you could do something like this:
```ruby
# simple case
context "with the same categories used in each month"
  it "correctly sums the categories by month"

# case that needs zero-months
context "with not all categories having a purchase in each month"
  it "correctly sums the categories by month, adding zero totals where appropriate"
```

steviekaye committed:
Clarified and extended tests

kfrn commented:
:tada: :tada: :tada:
